### DISCUSSION TITLE
Duplicate `WebSocketException` class in starlette and websockets

### DISCUSSION THREAD

**Discussion by ukalwa** [Kludex/uvicorn#2080] [Potential Issue]
Hello,

It looks like there is a websocket exception base class in both starlette and websockets. Should they be treated as separate. I ran into an issue where I am catching all websocket exceptions (including ConnectionClosedError) raised in a fastapi endpoint and using `startlette.exceptions.WebSocketException` didn't work. I had to use `websockets.exceptions.WebSocketException` to catch them.

Should I be using `websockets.exceptions` or `starlette.exceptions`? FastAPI exports starlette.exceptions but they don't seem to work in this case.

----
Here is a simple contrived example:

```python
from fastapi import FastAPI, WebSocket, WebSocketException, WebSocketDisconnect
import websockets.exceptions as ex
from traceback import print_exc
import asyncio

app = FastAPI()

@app.websocket("/ws/test")
async def test_ws_endpoint(websocket: WebSocket):
    await websocket.accept()
    try:
        while True:
             await websocket.send_json("test")
             await asyncio.sleep(10)
    except (WebSocketDisconnect, WebSocketException) as e:
        # Should get the trace here but it doesn't
        print_exc()
    except (ex.WebSocketException) as e:
        # Instead, the trace is printed here
        print_exc()
    finally:
        await websocket.close()
```

Here is the stack trace:
```
Traceback (most recent call last):
  File "/home/test/app.py", line 11, in test_ws_endpoint
    await websocket.send_json(None)
  File "/home/test/.venv/lib/python3.10/site-packages/starlette/websockets.py", line 173, in send_json
    await self.send({"type": "websocket.send", "text": text})
  File "/home/test/.venv/lib/python3.10/site-packages/starlette/websockets.py", line 85, in send
    await self._send(message)
  File "/home/test/.venv/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 65, in sender
    await send(message)
  File "/home/test/.venv/lib/python3.10/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 311, in asgi_send
    await self.send(data)  # type: ignore[arg-type]
  File "/home/test/.venv/lib/python3.10/site-packages/websockets/legacy/protocol.py", line 635, in send
    await self.ensure_open()
  File "/home/test/.venv/lib/python3.10/site-packages/websockets/legacy/protocol.py", line 944, in ensure_open
    raise self.connection_closed_exc()
websockets.exceptions.ConnectionClosedError: received 1005 (no status code [internal]); then sent 1005 (no status code [internal])
```

To reproduce, the client connects to the socket and then disconnects.

**Comment by Kludex** [comment]
What uvicorn version are you using?

**Reply by ukalwa** [reply]
Running uvicorn 0.20.0 with CPython 3.10.6 on Linux

**Reply by Kludex** [reply]
Does it reproduces with uvicorn 0.19.0?

**Reply by ukalwa** [reply]
Yes, it seems to happen in 0.19.0 as well

**Reply by Kludex** [reply]
I think I know what the problem is... Can you please check if it reproduces with `wsproto` instead of `websockets`?

You'll need to add `--ws wsproto` on the `uvicorn` command and `install `wsproto` (`pip install wsproto`).

**Reply by ukalwa** [reply]
Seems to be running into the same issue with the exceptions now coming from `wsproto`

`wsproto.utilities.LocalProtocolError: Event CloseConnection(code=1000, reason='') cannot be sent in state ConnectionState.CLOSED.`

**Reply by Kludex** [reply]
On uvicorn 0.20.0?

**Reply by ukalwa** [reply]
Yes, I just checked and it also happens on 0.19.0

**Reply by ukalwa** [reply]
FWIW, I think changing the [base ws exception](https://github.com/encode/starlette/blob/master/starlette/exceptions.py#L26) to inherit from base exception of `wsproto` or `websockets` instead of `Exception` might fix the issue?

**Reply by ukalwa** [reply]
Base exception in `websockets` is defined as:

```python
# websockets/exceptions.py#68
class WebSocketException(Exception):
    """
    Base class for all exceptions defined by websockets.

    """
```

And in `wsproto`:
```python
# wsproto/utilities.py#21
class ProtocolError(Exception):
    pass
```

**Reply by Kludex** [reply]
> Yes, I just checked and it also happens on 0.19.0

Yeah, but I thought I had fixed this on 0.20.0 for wsproto.

The issue is not on starlette, it's on uvicorn.

**Reply by ukalwa** [reply]
Maybe it's a different error(?) In my case, it is a SSE (server sent event), the api sends logs from the server side to connected clients periodically until the client disconnects.
